File: sdk/node/src/bot.ts
/**
 * Agent Poker SDK — Bot Class
 *
 * Handles the entire game loop: poll for state, check if it's your turn,
 * call your strategy, submit actions, and handle errors/reconnection.
 *
 * Usage:
 *   const bot = new Bot(client, {
 *     strategy: (state) => {
 *       if (state.availableActions.includes('check')) return { action: 'check' };
 *       return { action: 'fold' };
 *     }
 *   });
 *   bot.start();
 */

import { EventEmitter } from 'events';
import { AgentPokerClient, AgentPokerClientError } from './client';
import { GameState, StrategyFunction, StrategyResult, StrategyLike, ActionType } from './types';

export interface BotOptions {
  /** Your strategy function — receives game state, returns an action */
  strategy: StrategyFunction;
  /** Poll interval in milliseconds (default: 2000) */
  pollInterval?: number;
  /** Auto-rejoin if disconnected from table (default: true) */
  autoRejoin?: boolean;
  /** Table ID to join (default: "main") */
  tableId?: string;
  /** Enable verbose console logging (default: false) */
  verbose?: boolean;
  /** Auto-join table on start (default: true) */
  autoJoin?: boolean;
}

export interface BotEvents {
  handStart: (state: GameState) => void;
  handEnd: (state: GameState) => void;
  myTurn: (state: GameState) => void;
  bust: (chips: number) => void;
  error: (error: Error) => void;
  action: (result: StrategyResult, state: GameState) => void;
  connected: () => void;
  disconnected: () => void;
}

export class Bot extends EventEmitter {
  private client: AgentPokerClient;
  private strategy: StrategyFunction;
  private pollInterval: number;
  private autoRejoin: boolean;
  private tableId: string;
  private verbose: boolean;
  private autoJoin: boolean;

  private running = false;
  private lastHandId: string | null = null;
  private lastPhase: string | null = null;
  private pollTimer: ReturnType<typeof setTimeout> | null = null;

  constructor(client: AgentPokerClient, options: BotOptions) {
    super();
    this.client = client;
    this.strategy = options.strategy;
    this.pollInterval = options.pollInterval ?? 2000;
    this.autoRejoin = options.autoRejoin ?? true;
    this.tableId = options.tableId ?? 'main';
    this.verbose = options.verbose ?? false;
    this.autoJoin = options.autoJoin ?? true;
  }

  // ─── Public API ──────────────────────────────────────────────────────────

  /** Start the bot — joins the table and begins the game loop */
  async start(): Promise<void> {
    if (this.running) return;
    this.running = true;

    this.log('🃏 Agent Poker Bot starting...');

    // Graceful shutdown on SIGINT
    const shutdown = () => this.stop();
    process.on('SIGINT', shutdown);
    process.on('SIGTERM', shutdown);

    // Join table
    if (this.autoJoin) {
      await this.joinTable();
    }

    // Start polling
    this.log(`📡 Polling every ${this.pollInterval}ms`);
    this.poll();
  }

  /** Stop the bot gracefully */
  stop(): void {
    if (!this.running) return;
    this.running = false;

    if (this.pollTimer) {
      clearTimeout(this.pollTimer);
      this.pollTimer = null;
    }

    this.log('👋 Bot stopped');
    this.emit('disconnected');
  }

  // ─── Internal ────────────────────────────────────────────────────────────

  private async joinTable(): Promise<void> {
    try {
      const result = await this.client.join(this.tableId);
      this.log(`✅ Joined table: ${result.tableId}`);
      this.emit('connected');
    } catch (err) {
      if (err instanceof AgentPokerClientError) {
        // Already at table is fine
        if (err.statusCode === 400 && err.response?.error?.includes('already')) {
          this.log('ℹ️  Already at table');
          this.emit('connected');
          return;
        }
      }
      throw err;
    }
  }

  private async poll(): Promise<void> {
    if (!this.running) return;

    try {
      const state = await this.client.state();
      await this.processState(state);
    } catch (err) {
      await this.handleError(err as Error);
    }

    // Schedule next poll
    if (this.running) {
      this.pollTimer = setTimeout(() => this.poll(), this.pollInterval);
    }
  }

  private async processState(state: GameState): Promise<void> {
    // Detect new hand
    if (state.handId && state.handId !== this.lastHandId) {
      if (this.lastHandId !== null) {
        this.emit('handEnd', state);
      }
      this.lastHandId = state.handId;
      this.lastPhase = state.phase;
      this.emit('handStart', state);
      this.logVerbose(`🂠 New hand: ${state.handId} | Cards: ${state.yourCards.join(' ')}`);
    }

    // Detect phase change
    if (state.phase !== this.lastPhase) {
      this.lastPhase = state.phase;
      this.logVerbose(`📋 Phase: ${state.phase} | Board: ${state.communityCards.join(' ') || '—'} | Pot: $${state.pot}`);
    }

    // Detect bust
    if (state.yourChips <= 0 && state.phase === 'waiting') {
      this.emit('bust', state.yourChips);
      this.log('💀 Busted! Chips: 0');
    }

    // Act if it's our turn
    if (state.isYourTurn && state.availableActions.length > 0) {
      this.emit('myTurn', state);
      await this.takeAction(state);
    }
  }

  private async takeAction(state: GameState): Promise<void> {
    try {
      // Call the strategy function
      const raw = await Promise.resolve(this.strategy(state));
      const decision: StrategyResult = {
        action: raw.action as ActionType,
        amount: raw.amount,
        chat: raw.chat,
      };

      // Validate the action is available
      if (!state.availableActions.includes(decision.action)) {
        this.log(`⚠️  Strategy returned "${decision.action}" but available actions are: ${state.availableActions.join(', ')}. Folding.`);
        await this.client.act({ action: 'fold' });
        return;
      }

      // Submit the action
      await this.client.act({
        action: decision.action,
        amount: decision.amount,
      });

      this.emit('action', decision, state);
      this.log(`🎯 ${decision.action.toUpperCase()}${decision.amount ? ` $${decision.amount}` : ''} | Pot: $${state.pot} | Cards: ${state.yourCards.join(' ')}`);

      // Send chat if the strategy included one
      if (decision.chat) {
        try {
          await this.client.chat(decision.chat);
          this.logVerbose(`💬 "${decision.chat}"`);
        } catch {
          // Chat errors are non-critical
        }
      }
    } catch (err) {
      this.log(`❌ Action failed: ${(err as Error).message}`);
      this.emit('error', err as Error);

      // Try to fold as a fallback
      try {
        if (state.availableActions.includes('fold')) {
          await this.client.act({ action: 'fold' });
          this.log('🔄 Fell back to fold');
        }
      } catch {
        // Nothing more we can do
      }
    }
  }

  private async handleError(err: Error): Promise<void> {
    if (err instanceof AgentPokerClientError) {
      // Not at a table — try to rejoin
      if (err.statusCode === 400 && err.response?.error?.includes('Not at a table') && this.autoRejoin) {
        this.log('🔄 Not at table, attempting to rejoin...');
        try {
          await this.joinTable();
        } catch (joinErr) {
          this.log(`❌ Rejoin failed: ${(joinErr as Error).message}`);
        }
        return;
      }

      // Rate limited — back off
      if (err.statusCode === 429) {
        this.log('⏳ Rate limited, backing off...');
        await this.sleep(5000);
        return;
      }
    }

    // Network errors — log but keep going
    if ((err as any).name === 'AbortError' || (err as any).code === 'ECONNREFUSED') {
      this.logVerbose(`🌐 Network error: ${err.message}`);
      return;
    }

    this.log(`❌ Error: ${err.message}`);
    this.emit('error', err);
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  private log(msg: string): void {
    const ts = new Date().toLocaleTimeString();
    console.log(`[${ts}] ${msg}`);
  }

  private logVerbose(msg: string): void {
    if (this.verbose) this.log(msg);
  }
}

File: sdk/python/agent_poker/bot.py
"""
Agent Poker SDK — Bot Class

Handles the entire game loop: poll for state, check if it's your turn,
call your strategy, submit actions, and handle errors/reconnection.

Usage:
    from agent_poker import Client, Bot

    client = Client("pk_live_...")

    def my_strategy(state):
        if "check" in state.available_actions:
            return {"action": "check"}
        return {"action": "fold"}

    bot = Bot(client, strategy=my_strategy)
    bot.run()
"""

from __future__ import annotations
import signal
import sys
import time
from datetime import datetime
from typing import Callable, Optional, Dict, Any, Union

from .client import Client, ClientError
from .types import GameState, StrategyResult


# Strategy function type: receives GameState, returns dict or StrategyResult
StrategyFn = Callable[[GameState], Union[Dict[str, Any], StrategyResult]]


class Bot:
    """
    Poker bot that runs the game loop automatically.

    Args:
        client: An authenticated Agent Poker client.
        strategy: Function that takes a GameState and returns an action dict.
        poll_interval: Seconds between state polls (default: 2.0).
        auto_rejoin: Rejoin table if disconnected (default: True).
        table_id: Table to join (default: "main").
        verbose: Enable detailed logging (default: False).
        auto_join: Auto-join table on start (default: True).
    """

    def __init__(
        self,
        client: Client,
        strategy: StrategyFn,
        poll_interval: float = 2.0,
        auto_rejoin: bool = True,
        table_id: str = "main",
        verbose: bool = False,
        auto_join: bool = True,
    ):
        self.client = client
        self.strategy = strategy
        self.poll_interval = poll_interval
        self.auto_rejoin = auto_rejoin
        self.table_id = table_id
        self.verbose = verbose
        self.auto_join = auto_join

        self._running = False
        self._last_hand_id: Optional[str] = None
        self._last_phase: Optional[str] = None

        # Event callbacks
        self._callbacks: Dict[str, list] = {
            "hand_start": [],
            "hand_end": [],
            "my_turn": [],
            "bust": [],
            "error": [],
            "action": [],
            "connected": [],
            "disconnected": [],
        }

    # ─── Events ───────────────────────────────────────────────────────────

    def on(self, event: str, callback: Callable) -> "Bot":
        """Register an event callback. Returns self for chaining."""
        if event in self._callbacks:
            self._callbacks[event].append(callback)
        return self

    def _emit(self, event: str, *args: Any) -> None:
        """Emit an event to all registered callbacks."""
        for cb in self._callbacks.get(event, []):
            try:
                cb(*args)
            except Exception as e:
                self._log(f"❌ Callback error ({event}): {e}")

    # ─── Public API ───────────────────────────────────────────────────────

    def run(self) -> None:
        """Start the bot — joins the table and runs the game loop. Blocks until stopped."""
        self._running = True
        self._log("🃏 Agent Poker Bot starting...")

        # Graceful shutdown
        def handle_signal(sig: int, frame: Any) -> None:
            self._log("\n👋 Shutting down...")
            self.stop()
            sys.exit(0)

        signal.signal(signal.SIGINT, handle_signal)
        signal.signal(signal.SIGTERM, handle_signal)

        # Join table
        if self.auto_join:
            self._join_table()

        # Game loop
        self._log(f"📡 Polling every {self.poll_interval}s")
        while self._running:
            try:
                state = self.client.state()
                self._process_state(state)
            except Exception as e:
                self._handle_error(e)

            time.sleep(self.poll_interval)

    def stop(self) -> None:
        """Stop the bot gracefully."""
        self._running = False
        self._log("👋 Bot stopped")
        self._emit("disconnected")

    # ─── Internal ─────────────────────────────────────────────────────────

    def _join_table(self) -> None:
        """Join the configured table."""
        try:
            result = self.client.join(self.table_id)
            self._log(f"✅ Joined table: {result.get('tableId', self.table_id)}")
            self._emit("connected")
        except ClientError as e:
            if "already" in str(e).lower():
                self._log("ℹ️  Already at table")
                self._emit("connected")
            else:
                raise

    def _process_state(self, state: GameState) -> None:
        """Process a game state update."""
        # Detect new hand
        if state.hand_id and state.hand_id != self._last_hand_id:
            if self._last_hand_id is not None:
                self._emit("hand_end", state)
            self._last_hand_id = state.hand_id
            self._last_phase = state.phase
            self._emit("hand_start", state)
            self._log_verbose(
                f"🂠 New hand: {state.hand_id} | Cards: {' '.join(state.your_cards)}"
            )

        # Detect phase change
        if state.phase != self._last_phase:
            self._last_phase = state.phase
            board = " ".join(state.community_cards) or "—"
            self._log_verbose(f"📋 Phase: {state.phase} | Board: {board} | Pot: ${state.pot}")

        # Detect bust
        if state.your_chips <= 0 and state.phase == "waiting":
            self._emit("bust", state.your_chips)
            self._log("💀 Busted! Chips: 0")

        # Act if it's our turn
        if state.is_your_turn and state.available_actions:
            self._emit("my_turn", state)
            self._take_action(state)

    def _take_action(self, state: GameState) -> None:
        """Call the strategy and submit an action."""
        try:
            # Call strategy
            result = self.strategy(state)

            # Normalize result to dict
            if isinstance(result, StrategyResult):
                decision = {
                    "action": result.action,
                    "amount": result.amount,
                    "chat": result.chat,
                }
            else:
                decision = result

            action = decision.get("action", "fold")
            amount = decision.get("amount")
            chat_text = decision.get("chat")

            # Validate action
            if action not in state.available_actions:
                self._log(
                    f"⚠️  Strategy returned '{action}' but available: "
                    f"{', '.join(state.available_actions)}. Folding."
                )
                self.client.act("fold")
                return

            # Submit action
            self.client.act(action, amount)
            self._emit("action", decision, state)

            amount_str = f" ${amount}" if amount else ""
            cards = " ".join(state.your_cards)
            self._log(f"🎯 {action.upper()}{amount_str} | Pot: ${state.pot} | Cards: {cards}")

            # Send chat if included
            if chat_text:
                try:
                    self.client.chat(chat_text)
                    self._log_verbose(f'💬 "{chat_text}"')
                except Exception:
                    pass  # Chat errors are non-critical

        except Exception as e:
            self._log(f"❌ Action failed: {e}")
            self._emit("error", e)

            # Fallback: try to fold
            try:
                if "fold" in state.available_actions:
                    self.client.act("fold")
                    self._log("🔄 Fell back to fold")
            except Exception:
                pass

    def _handle_error(self, err: Exception) -> None:
        """Handle errors during polling."""
        if isinstance(err, ClientError):
            # Not at a table — try to rejoin
            if err.status_code == 400 and "not at a table" in str(err).lower() and self.auto_rejoin:
                self._log("🔄 Not at table, attempting to rejoin...")
                try:
                    self._join_table()
                except Exception as e:
                    self._log(f"❌ Rejoin failed: {e}")
                return

            # Rate limited — back off
            if err.status_code == 429:
                self._log("⏳ Rate limited, backing off...")
                time.sleep(5)
                return

        self._log(f"❌ Error: {err}")
        self._emit("error", err)

    def _log(self, msg: str) -> None:
        """Print a timestamped log message."""
        ts = datetime.now().strftime("%H:%M:%S")
        print(f"[{ts}] {msg}")

    def _log_verbose(self, msg: str) -> None:
        """Print a log message only in verbose mode."""
        if self.verbose:
            self._log(msg)
